### DISCUSSION TITLE
Should we update the entire imported react-pro-metronome library to typescript?

### DISCUSSION THREAD

**Discussion by dcallus** [fillmaster/fillMaster#41] [Polls]
The original library doesn't look like it's going to be updated and even if it were, we need to have it modified for various issues.

We will be adding options for additional sounds, amongst other things, so it might be good to clean it up before we go any further.

I propose that we better integrate this part of the app, by 
a) converting it to typescript. 
b) not having it imported like a library, i.e. no source folder etc. This way we can move the tests out into the route src folder and get them up and running again.

I definitely think we should leave it state based though. It works perfectly and I don't see the need to convert to functional react. It may not be worth the effort. It may help keep things consistent if we do, but may be easier to add patches if the original library changes. It's kind of nice having it quite isolated and separate as it works so well, but inevitably it will become more integrated over time. I probably need some better folder structuring moving forward.

I'm particularly interested in your opinion [chrisd1999](https://github.com/chrisd1999) as you've worked a little in this part of the app.

**Comment by kristersd** [comment]
@dcallus I think, we should create an separate type-safe functional component ourselves, and link the original authors. As there is only one component, which manages the logic, we could split the functionality to be more robust. I can work on this particular implementation, if that may be hard to others. This way we can test it thoroughly, and not rely on their class component implementation.

**Reply by dcallus** [reply]
Wow. Of course. If you could do this that would be absolutely amazing. Thank you so much  🥇  👍  💯 !

There are only two stipulations however before proceeding:

1. It must use howlerjs (or any other library that implements the 'Web Audio API' or just Web Audio API itself) as the original library (react-pro-metronome) does, as it's designed for audio stability. No point in re-inventing the wheel here. As it's a metronome (and a metronome for drummers at that) the timing needs to be impeccable and rock solid. Using Web Audio API is the done thing from my research.

2. I'd like it to take the same props please (obviously more is fine :) ). But I really like being able to just make a string of e.g. '100200200' with the patternMaker to choose where the metronome beats and sub-beats lie. It's a beautiful abstraction that makes it easier for more intermediate devs like myself to interface with metronome. It also means we won't need to change the code outside of the metronome too much, if it all.

Would that be okay?

I am open of course if you have a really good reason to do otherwise.

**Reply by dcallus** [reply]
Hope this didn't come off as rude btw. I'm just trying to be organised, lol :) 

Your help and experience is super appreciated. :)

**Reply by kristersd** [reply]
@dcallus I refactored to use hook instead of component, could you check/test if it does work as expected? Branch https://github.com/fillmaster/fillMaster/tree/useprometronome-hook

**Reply by dcallus** [reply]
The hook seems to work really well. It took me a few reads through but I pretty much get what's going on now. To me it's clearer than the previous library.

Also from first use it seems a little more stable on start up and when navigating to and from the app etc. Which is nice.

One thing I'm not 100% sure about though (and I wasn't actually sure about on the previous implementation either) is this, is Howler actually doing anything except playing the files? What's keeping the time? I'm finding it hard to grasp this.

Here is an article I read a while ago before starting this project https://blog.paul.cx/post/metronome/ and I also read the linked post at the start of the third paragraph. Was really interesting stuff about creating stable metronomes. Things have perhaps moved on since then and less code is necessary to achieve the same results, but I'm honestly not sure if we are implementing this correctly.

Part of me always thought that maybe Howler was doing a lot of this magic and it has abstracted the Web Audio API really well.  But now I'm starting to think (now that the code is clear to read) that the timing is actually just done in plain javascript (e.g. pro-metronome and your new hook). I've not managed to confirm this because for the life of me I can't figure out where this is happening, despite how simple the code now looks. lol.

I mean the metronome SOUNDS pretty damn stable to me. But actually, now that I'm testing it and flicking to different tabs on my browser it is (I think) becoming a little unstable, I'd need to do some tests, it's quite hard to tell tbh. But if you could clear that up for me, where the logic for the timing is happening that'd be appreciated. 

Then we can move forward with a PR as I'd like to make a couple of in-line comments if possible.

Thanks again :)

**Reply by dcallus** [reply]
Yeah. So unfortunately I've found some stability issues that have been introduced. If you play the metronome and go between quarter note and eight notes for example there is a slight lag before it updates. The previous implementation you could flick between any subdivision as often as you like (whilst it's playing) and it would remain perfectly in time. Also I went back and checked previous implementation and it seems pretty stable when changing tabs etc and just sounds overall more stable. Perhaps related to what I've mentioned above, but I'm not sure.

**Reply by kristersd** [reply]
@dcallus Ok, gonna check it out. It may be related when the useEffect clean up function run. Thanks for feedback.

**Reply by dcallus** [reply]
Somehow didn't see this message.. Nice one.. Thanks for looking into it. :) Let me know how it goes.

**Reply by kristersd** [reply]
@dcallus I have not got the time to check this issue out, hopefully I can manage to do it this weekend. 😬

**Reply by dcallus** [reply]
No worries @chrisd1999 - Your help is always appreciated. Whenever you can, this isn't an urgent feature but it's definitely a nice to have and will definitely help moving forward. We can progress fine for now though in other areas :) 

Also if it's possible before starting, would you be able to tell me if you think that Howler is used (in the original pro-metronome code or your own) at all for stability? For me it looks like it only is used to play the audio.

If it doesn't factor in stability could you perhaps look into the post I mentioned above (https://blog.paul.cx/post/metronome/) and see if it's possible to implement the recommended WebAudioAPI into the metronome's core? along with any appropriate cacheing etc mentioned in the article, or even better, see if Howler can abstract this work as it implements WebAudioAPI.

Thanks again

**Comment by dcallus** [comment]
Apparently useMemo is a fair amount slower than useState, especially on initial render, which may hint at why it is slow when changing.

I've not had any issues with the rest of the app slowing down to be honest, so could be safe just to use useState as to be honest, the metronome must take priority over all else. I'll attempt to change it and see if it works.

Note 100% if this is relevant, but thought it was worth noting:
https://canimerge.com/should-you-really-use-usememo-in-react-lets-find-out/

edit: nope. that wasn't the issue.
